### ISSUE TITLE
Problem with glBindAttribLocation with es_200

### ISSUE THREAD

**Issue by NicoKiaru** [elect86/jogl-samples#41]
Hello, 

I have a hard time trying to customize shaders with es_200 (I don't want to go above for compatibility reasons). In fact I noticed that changing [this line](https://github.com/java-opengl-labs/jogl-samples/blob/master/jogl-samples/src/tests/es_200/Es_200_draw_elements.java#L108) of `Es_200_draw_elements.java` from:

`gl2es2.glBindAttribLocation(programName, Semantic.Attr.POSITION, "position");`

to 

`gl2es2.glBindAttribLocation(programName, Semantic.Attr.POSITION, "pouetpouet");`

or whatever gibberish as string do not have any effect : both are working well. That's weird. I do not understand why. 

Something also weird, if I change the vertex shader `flat-color.vert` from:

```uniform mat4 mvp;
attribute highp vec4 position;
void main()
{
    gl_Position = mvp * position;
}
```

to 
```
uniform mat4 mvp;
attribute highp vec4 positionwhatever;
void main()
{	
    gl_Position = mvp * positionwhatever;
}
```

This works as well (keep "position" in the java code)... But how opengl is able to find and link the correct position variable ?... I'm really confused. 

On the contrary the link with Uniforms is working as expected and consistent : for instance I need to change "mvp" both in the shader and in the java code. But what's the deal with `glBindAttribLocation`?

Thanks for any help...

**Comment by elect86** [comment]
Hi nico

one reason is that `Semantic.Attr.POSITION` is actually `0` and the driver is probably falling back to assign the vertex attributes to integers as they are found by the compiler while parsing the shader.

One call that might actually helps you clearing what's going on is [`glGetAttribLocation`](https://www.khronos.org/registry/OpenGL-Refpages/gl4/html/glGetAttribLocation.xhtml).

`glGetAttribLocation(programName, "position")` should returns indeed `0`, while `glGetAttribLocation(programName, "pouetpouet")` instead `-1`, that stays for invalid.

Please note that the location are taken into account **only** after the program linking..!

Let me know any furthers progressions on the matter, I am also interested.


Ps: I improved the layout of your post, I hope you don't mind :)
Pps: are you on Android? Would you be interested into more samples?

**Comment by NicoKiaru** [comment]
Hi Giuseppe, 

Thanks a lot for your answer (and also for the editing -> I could learn what I should have done). 

I managed to solve the issue ! As you wrote:

> Please note that the location are taken into account only after the program linking..!

But in most of the samples of this repository, the link is made after the call to 'glGetAttribLocation' - see for instance L108 and L110  [here](https://github.com/java-opengl-labs/jogl-samples/blob/master/jogl-samples/src/tests/es_200/Es_200_draw_elements.java#L108).

I switched it in my code and now the behaviour is as expected. I believe you may want to switch them in your code as well. Tell me!

pps : I'm not on Android and personally I do not need more samples. Most of the problems I encounter are just related to installing properly JOGL on several machines and to know which versions are supported. This repository is nice because I could check until which version I could go on my laptop (fyi : not above es_200, and I even had to remove the command 'gl2es2.glClearDepthf(1f);'). 

-> I'm keeping an eye on related projects implemented with Maven.

**Comment by elect86** [comment]
Sorry, let me rephrase, the vertex attributes get bounded **at** the program linking.

Therefore, if you want to bind a vertex attribute to a specific location index with `glBindAttribLocation`, you have to do before the program linking.
Otherwise if you don't care and you are inclined to use whatever location the compiler decides itself, you will call `glGetAttribLocation` after the program linking.

What problems did you encounter? The setup is pretty simple:

- add the `gluegen-rt.jar` and `jogl-all.jar`
- add a folder with extracted natives or leave the native jars in the same folder of gluegen and jogl

I'm going to write sooner or later a small utility to detect the highest available version of opengl and list all the capabilities.. maybe this might interest you?

If you are on desktop, you should use plain gl, es-gl is for mobile or testing mobile on desktop

Ps: to format code you need the backtick `, not the normal '

I know almost nothing about Maven, but I had some fun with gradle recently, may it be an alternative for you? If yes, I can help you with that

**Comment by NicoKiaru** [comment]
OK, so I've learned something today ;)

I was definitely confused because I didn't get the difference between `glBindAttribLocation` and `glGetAttribLocation`...

Everything is working fine now, and I could add whatever parameter I want in the shaders. 

> If you are on desktop, you should use plain gl, es-gl is for mobile or testing mobile on desktop

Indeed, I didn't get the differences. I moved everything to plain gl, and hopefully this was very straightforward.

I think for OpenGL, I'm not so interested in knowing the highest version available. I aim for using a version for the largest number of people. To be more specific, I develop a plugin for ImageJ which is an image analysis software (https://imagej.net/ImageJ2). And the reason I'm interested  in maven is because apparently that's what they advise for ImageJ plugin: (https://imagej.net/Distribution).

Thanks a lot for your help, and by the way, I began this discussion as an issue because I didn't find a sort of generic forum to discuss my problem on this repository. I'm quite new to GitHub, so please, if there is a more appropriate place to discuss this sort of problem, tell me!

**Comment by elect86** [comment]
> OK, so I've learned something today ;)
> Everything is working fine now, and I could add whatever parameter I want in the shaders.

Nice :p

> I aim for using a version for the largest number of people.

Ok, then you should aim at OpenGL 3 (released on [2008](https://en.wikipedia.org/wiki/OpenGL#OpenGL_3.3)), that should be more than enough

As far as I know, Gradle is 100% Maven compatible, but use Maven if they say so (maybe you could ask them about Maven/Gradle)

This place is ok as far as the question regards the code/project itself

If you need generic opengl help instead, consider [StackOverflow](http://stackoverflow.com/) or jogamp [forums](http://forum.jogamp.org/)
